xds/src/main/java/io/grpc/xds/internal/certprovider/FileWatcherCertificateProvider.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.internal.certprovider;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.InternalLogId;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.internal.TimeProvider;
+import io.grpc.xds.internal.sds.trust.CertificateUtils;
+
+import java.io.ByteArrayInputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.FileTime;
+import java.security.PrivateKey;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/** Implementation of {@link CertificateProvider} for file watching cert provider. */
+final class FileWatcherCertificateProvider extends CertificateProvider {
+  private static final Logger logger =
+      Logger.getLogger(FileWatcherCertificateProvider.class.getName());
+
+  private final SynchronizationContext syncContext;
+  private final ScheduledExecutorService scheduledExecutorService;
+  private final TimeProvider timeProvider;
+  private final Path certFile;
+  private final Path keyFile;
+  private final Path trustFile;
+  private final long refreshIntervalInSeconds;
+  @VisibleForTesting SynchronizationContext.ScheduledHandle scheduledHandle;
+  private FileTime lastModifiedTimeCert;
+  private FileTime lastModifiedTimeKey;
+  private FileTime lastModifiedTimeRoot;
+
+  FileWatcherCertificateProvider(
+      DistributorWatcher watcher,
+      boolean notifyCertUpdates,
+      String certFile,
```

Comment:
As mentioned offline this is going to eventually replace `DynamicReloadingCertificateProvider` since the design now calls for a generic file-watcher plugin instead of assuming a symlink based provider. So we don't assume a common directory.